Twine "decision" nodes

grahamearley opened the pull request:
Sometimes the game creator might want a player to be able to move to the next Twine node without actually doing anything in the scene --- instead, the player is making some sort of decision in their head (i.e. "Today I'm going to rebel in the workhouse!").

### This PR introduces the idea of "Decision nodes"
You can mark any `TwineNode` as a decision node in the inspector GUI. 

**You can also designate a node as a decision node by tagging the node with `prairie_decision` in the Twine web interface.** When a node is designated as a decision node, the GUI checkbox will be marked, *and* the tag will be added to the node.

***

### How it works in the in-game GUI

When the user reaches a decision node, the GUI will look like this:

![](https://media.giphy.com/media/26gsvHTwLdn1whfxK/giphy.gif)

You can press `Q` to expand the list of titles of child nodes you can visit. When you expand this list, you can scroll through it by pressing `Q` (or close the list by pressing `E`). Then you can select your desired node to move to by pressing `ENTER` on it.

grahamearley committed:
add `isDecisionNode` field to TwineNodes, add GUI for toggling

grahamearley committed:
Add GUI for progressing story if node is a decision node

grahamearley committed:
Ensure that the TwineNode is updated if GUI changes

(otherwise the isDecisionNode value gets reset when the user presses
Play)

grahamearley committed:
Add GUI list of decisions to make from a decision node, with SOFT activation of other nodes

soft activation means there is no INTERACTOR attached, so we don’t call
the objects to trigger!

grahamearley committed:
Set child options GUI to draw in same frame as all Twine info

and change which keys control the opening and closing

grahamearley committed:
Remove "soft activate" method, and instead find a reference to the interaction to use normal activate method

grahamearley committed:
Link up the "is decision node" checkbox and the "prairie_decision" tag

so both appear if the node is a decision node and both disappear if
not. Also check the tag when the node is created

grahamearley commented:
Also, the keys that control the GUI, as well as all the text, is up for discussion :)

cpimhoff commented on Project/Assets/Editor/PrairieGUI/TwineNodeInspector.cs:
```diff
@@ -20,5 +26,34 @@ public override void OnInspectorGUI ()
+		// For the node to be a decision node, we'll say it has 
+		//	to have a "prairie_decision" tag, AND it has to have 
+		//	the isDecisionNode checkbox checked.
+		bool hasDecisionTag = node.tags.Contains (TwineNode.PRAIRIE_DECISION_TAG);
```
I don't understand. A node has to both have this tag _and_ a user needs to manually check a box on the node itself? 

Could we just make it a checkbox on the node, and have the importer set it to `true` if the tag is detected on import? This two-step setup would certainly throw me, and seems to complicate the logic in the GUI to boot.

grahamearley replied:
For the record, it's not a two-step setup. It's just that there are two possible ways to set a node as a decision node (you can either check the box, or add the "prairie_decision" tag)

grahamearley replied:
But I'm still going to streamline it since we're going to get rid of tags in the TwineNode inspector

cpimhoff commented on Project/Assets/Prairie/Framework/Script/Twine/TwineNode.cs:
```diff
@@ -15,14 +19,43 @@ public class TwineNode : MonoBehaviour {
+	{
+		// Upon creation of this node, ensure that it is a decision node if it has
+		//	the decision tag:
+		this.isDecisionNode = this.tags.Contains (PRAIRIE_DECISION_TAG);
```
Now I'm extra confused. Why did we even bother having `isDecisionNode` be a property if we're just going to reset it when we launch the game?

grahamearley replied:
My bad -- I was thinking the `Start()` method gets called on creation, not on play.

Will add this logic to the Twine importer instead.

cpimhoff commented on Project/Assets/Prairie/Framework/Script/Twine/TwineNode.cs:
```diff
@@ -15,14 +19,43 @@ public class TwineNode : MonoBehaviour {
+				this.selectedOptionIndex = (this.selectedOptionIndex + 1) % (children.Length);
+			} else if (this.isOptionsGuiOpen && (Input.GetKeyDown (KeyCode.KeypadEnter) || Input.GetKeyDown (KeyCode.Return))) {
+				// Press ENTER or RETURN to select that child
+				this.ActivateChildAtIndex (selectedOptionIndex);
```
I think we should put a check in and to ensure this is the currently activated node, before we risk modifying any state based on user input. I don't know - this just worries me since update gets run on every twine node (not just the active one).

I'm pretty sure that any node with `.enabled = false` won't run their `Update()` method (so this isn't really a problem), but a sanity check would go a long way in making it clearer.

If we later decide that twine node's should have a `isCurrentNode` property (instead of piggybacking off of `.enabled`, which has side effects), a bug would be introduced here.

grahamearley replied:
Good call -- done

cpimhoff commented on Project/Assets/Prairie/Framework/Script/Twine/TwineNode.cs:
```diff
@@ -31,24 +64,44 @@ public void OnGUI()
 
 			float frameWidth = Screen.width / 3;
-			Rect frame = new Rect (10, 10, frameWidth, Screen.height);
+			Rect frame = new Rect (10, 10, frameWidth, Screen.height / 2);
```
It'd be great to introduce max sizes here - on big screens, or projectors (which we'll be demoing on), half of the screen is a huge amount more space than really needed.

grahamearley replied:
✅

grahamearley replied:
Here's a sneak preview of the new resizing:
![](https://media.giphy.com/media/26xBzWUZtbYJvYo2Q/giphy.gif)

cpimhoff commented on Project/Assets/Prairie/Framework/Script/Twine/TwineNode.cs:
```diff
@@ -31,24 +64,44 @@ public void OnGUI()
+				if (!isOptionsGuiOpen) {
+					GUILayout.Box ("Press Q to progress in the story...", decisionHintStyle);
+				} else {
+					GUILayout.Box ("Press Q to scroll, E to close, ENTER to choose", decisionHintStyle);
```
I'm not crazy about this input style - but I have no alternative idea to suggest. All our best keyboard keys are already bound to things. 

:(

cpimhoff commented on Project/Assets/Prairie/Framework/Script/Twine/TwineNode.cs:
```diff
@@ -31,24 +64,44 @@ public void OnGUI()
+				GUIStyle optionButtonStyle = new GUIStyle (GUI.skin.button);
+				optionButtonStyle.fontStyle = FontStyle.Italic;
+				optionButtonStyle.wordWrap = true;
+				selectedOptionIndex = GUILayout.SelectionGrid(selectedOptionIndex, childrenNames, 1, optionButtonStyle);
```
Is there a way in this `SelectionGrid` method (I think by modifying the style param) to differentiate the currently highlighted (item at `selectedOptionIndex`) a bit more from the rest? Right now they look kinda similar.

grahamearley replied:
Done. I grayed out the text for the inactive selection. Looks like this:

![](https://media.giphy.com/media/26gs7IcbWC7xF9ALC/giphy.gif)

grahamearley replied:
I think that makes it pretty clear what you're selecting. However, we can also modify the background texture -- but we'd have to make custom textures to do that, so maybe later

cpimhoff reviewed (changes requested):
This is a great PR, and I think this feature is going to lead to a lot more interesting Twine stories.

I don't yet understand the interplay between the tag on the twine node and the node's `isDecisionNode` parameter. I think this should be simplified to one or the other, which will in turn probably help with the GUI's current complexity in dealing with syncing the two.

Should the `TwineNodeGUI` even present the list of tags to the user? I think the importer should just use these. There's no reason I can think of for a user to edit them after import.

I wrote a few other comments, but I'm happy to pull this in before those are resolved.

grahamearley committed:
Merge remote-tracking branch 'origin/master' into merge-gui-changes

grahamearley committed:
Merge branch 'merge-gui-changes' into graham/twine-decisions

grahamearley committed:
Moved prairie_decision tag check to Twine importer

Also removed decision status check on tag update

grahamearley committed:
Ensure that TwineNode Update() is only called if enabled

this lets us easily switch from using `.enabled` to `.isCurrentNode` or some other boolean field we could create.

grahamearley committed:
Set min width/height for the TwineNode GUI so it doesn’t get too big

grahamearley committed:
In decision selection grid, highlight selected choice with white text

and give the unselected node darker gray text

cpimhoff commented on Project/Assets/Editor/CustomInspectors/TwineNodeInspector.cs:
```diff
@@ -2,15 +2,21 @@
 	{
 		node = (TwineNode)target;
+
+		bool wasDecisionNode = node.isDecisionNode;
```
Where is this variable used?

grahamearley replied:
Not used! I just removed it.

cpimhoff commented on Project/Assets/Prairie/Framework/Script/Twine/TwineNode.cs:
```diff
@@ -15,40 +17,90 @@ public class TwineNode : MonoBehaviour {
+				// Set highlighted button to have green text (this state is called `onNormal`):
+				optionButtonStyle.onNormal.textColor = Color.white;
+				// Set non-highlighted buttons to have grayed out text (state is called `normal`)
+				optionButtonStyle.normal.textColor = Color.gray;
```
Good change. LITERALLY CAN NOT believe Unity decided 'normal' and 'onNormal' were good names for these states.

grahamearley replied:
yes. wat

cpimhoff commented on Project/Assets/Prairie/Framework/Script/Twine/TwineNode.cs:
```diff
@@ -77,11 +129,31 @@ public void Activate(GameObject interactor)
+
+			// Now activate the child using this interactor!
+			TwineNode child = this.children [index].GetComponent<TwineNode> ();
+			child.Activate (interactorObject);
```
You _might_ be able to get away with passing null here. But I like how we're being safe!

grahamearley replied:
✋

cpimhoff replied:
🤚

cpimhoff reviewed (approved):
All good here!
Get it in there!

grahamearley committed:
Remove unused variable

mjavaly reviewed (approved):
Like all these changes! Preparing to merge.
